bridge/client/app/_guards/pending-changes.guard.spec.ts

```diff
@@ -0,0 +1,80 @@
+import { TestBed } from '@angular/core/testing';
+
+import { PendingChangesComponent, PendingChangesGuard } from './pending-changes.guard';
+import { Observable, Subject } from 'rxjs';
+
+type NotificationState = null | 'unsaved';
+class MockComponent implements PendingChangesComponent {
+  private formUnsavedSubject = new Subject<boolean>();
+  returnValue: boolean | Observable<boolean> = true;
+  notificationState: NotificationState = null;
+
+  formTouched(): void {
+    this.returnValue = this.formUnsavedSubject.asObservable();
+  }
+
+  saveForm(): void {
+    this.formUnsavedSubject.next(true);
+  }
+
+  rejectChanges(): void {
+    this.formUnsavedSubject.next(false);
+  }
+
+  canDeactivate(): boolean | Observable<boolean> {
+    return this.returnValue;
+  }
+
+  showNotification($event?: BeforeUnloadEvent): void {
+    this.notificationState = 'unsaved';
+  }
+
+  hideNotification(): void {
+    this.notificationState = null;
+  }
+}
+
+describe('PendingChangesGuard', () => {
+  let mockComponent: MockComponent;
+  let guard: PendingChangesGuard;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [MockComponent],
+    });
+    guard = TestBed.inject(PendingChangesGuard);
+    mockComponent = TestBed.get(MockComponent);
+  });
+
+  it('expect guard to instantiate', () => {
+    expect(guard).toBeTruthy();
+  });
+
+  it('can route if unguarded', () => {
+    expect(guard.canDeactivate(mockComponent)).toBeTruthy();
+  });
+
+  it('will route if guarded and user accepted the dialog', () => {
+    // Mock the behavior of the GuardedComponent:
+    mockComponent.formTouched();
+    const canDeactivate$ = <Observable<boolean>>guard.canDeactivate(mockComponent);
+    canDeactivate$.subscribe((deactivate) => {
+      // This is the real test!
+      expect(deactivate).toBeTruthy();
```

Comment:
Is this even called? Because this is asynchronous I would expect that the test runs until `mockComponent.saveForm()` and finishes without waiting for the subscribe without the use of `waitForAsync`

Reply:
This simulates the user deciding to leave the page.

Reply:
I'm talking about the asynchronous `subscribe`. `subscribe` would be called right after `mockComponent.saveForm();`, but after this, Jest will tell that the test succeeded, no matter what's inside the `subscribe`, because it does not wait for ongoing asynchronous tasks.
No matter if `expect(deactivate).toBeTruthy();` or `expect(deactivate).toBeFalsy();` is called inside the subscribe, the test will pass.

Reply:
You are right. I added `waitForAsync`